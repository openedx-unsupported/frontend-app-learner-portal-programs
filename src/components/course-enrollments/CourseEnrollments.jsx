import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { breakpoints, StatusAlert } from '@edx/paragon';
import { AppContext } from '@edx/frontend-platform/react';
import { faExclamationTriangle, faCheckCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import MediaQuery from 'react-responsive';
import { LoadingSpinner } from '../loading-spinner';

import CourseSection from './CourseSection';
import {
  InProgressCourseCard,
  UpcomingCourseCard,
  CompletedCourseCard,
} from './course-cards';

import * as selectors from './data/selectors';
import * as actions from './data/actions';

export class CourseEnrollments extends Component {
  componentDidMount() {
    const {
      pageContext: {
        programUUID,
      },
    } = this.context;
    const { fetchCourseEnrollments } = this.props;
    const options = {};
    if (programUUID) {
      options.programUUID = programUUID;
    }
    fetchCourseEnrollments(options);
  }

  componentWillUnmount() {
    const { clearCourseEnrollments } = this.props;
    clearCourseEnrollments();
  }

  hasCourseRunsWithStatus = (status) => {
    const { courseRuns } = this.props;
    return courseRuns && courseRuns[status] && courseRuns[status].length > 0;
  };

  hasCourseRuns = () => (
    this.hasCourseRunsWithStatus('completed')
    || this.hasCourseRunsWithStatus('in_progress')
    || this.hasCourseRunsWithStatus('upcoming')
  );

  renderError = () => (
    <StatusAlert
      alertType="danger"
      dialog={(
        <div className="d-flex">
          <div>
            <FontAwesomeIcon className="mr-2" icon={faExclamationTriangle} />
          </div>
          <div>
            An error occurred while retrieving your course enrollments. Please try again.
          </div>
        </div>
      )}
      dismissible={false}
      open
    />
  );

  renderMarkCourseCompleteSuccessAlert = () => {
    const { modifyIsMarkCourseCompleteSuccess } = this.props;
    return (
      <StatusAlert
        alertType="success"
        dialog={(
          <div className="d-flex">
            <div>
              <FontAwesomeIcon className="mr-2" icon={faCheckCircle} />
            </div>
            <div>
              Your course was marked as complete.
            </div>
          </div>
        )}
        onClose={() => {
          modifyIsMarkCourseCompleteSuccess({ isSuccess: false });
        }}
        open
      />
    );
  };

  render() {
    const {
      children,
      courseRuns,
      isLoading,
      error,
      sidebarComponent,
      isMarkCourseCompleteSuccess,
    } = this.props;

    if (isLoading) {
      return <LoadingSpinner screenReaderText="loading course enrollments" />;
    } if (error) {
      return this.renderError();
    }

    return (
      <>
        {isMarkCourseCompleteSuccess && this.renderMarkCourseCompleteSuccessAlert()}
        {/*
          Only render children if there are no course runs.
          This allows the parent component to customize what
          gets displayed if the user does not have any course runs.
        */}
        {!this.hasCourseRuns() && children}
        <CourseSection
          title="My courses in progress"
          component={InProgressCourseCard}
          courseRuns={courseRuns.in_progress}
        />
        <MediaQuery minWidth={breakpoints.large.minWidth}>
          {matches => !matches && (
            <div className="mb-5">
              {sidebarComponent}
            </div>
          )}
        </MediaQuery>
        <CourseSection
          title="Upcoming courses"
          component={UpcomingCourseCard}
          courseRuns={courseRuns.upcoming}
        />
        <CourseSection
          title="Completed courses"
          component={CompletedCourseCard}
          courseRuns={courseRuns.completed}
        />
      </>
    );
  }
}

const mapStateToProps = state => ({
  courseRuns: selectors.getCourseRunsByStatus(state),
  isLoading: selectors.getIsLoading(state),
  error: selectors.getError(state),
  isMarkCourseCompleteSuccess: selectors.getIsMarkCourseCompleteSuccess(state),
});

const mapDispatchToProps = dispatch => ({
  fetchCourseEnrollments: (options) => {
    dispatch(actions.fetchCourseEnrollments(options));
  },
  clearCourseEnrollments: () => {
    dispatch(actions.clearCourseEnrollments());
  },
  modifyIsMarkCourseCompleteSuccess: (options) => {
    dispatch(actions.updateIsMarkCourseCompleteSuccess(options));
  },
});

CourseEnrollments.contextType = AppContext;

CourseEnrollments.propTypes = {
  fetchCourseEnrollments: PropTypes.func.isRequired,
  clearCourseEnrollments: PropTypes.func.isRequired,
  courseRuns: PropTypes.shape({
    in_progress: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    upcoming: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    completed: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  sidebarComponent: PropTypes.element.isRequired,
  isMarkCourseCompleteSuccess: PropTypes.bool.isRequired,
  modifyIsMarkCourseCompleteSuccess: PropTypes.func.isRequired,
  error: PropTypes.instanceOf(Error),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

CourseEnrollments.defaultProps = {
  error: null,
  children: null,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CourseEnrollments);
