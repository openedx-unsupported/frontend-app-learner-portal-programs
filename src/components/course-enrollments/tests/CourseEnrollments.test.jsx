/* eslint-disable react/jsx-no-constructed-context-values */
import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { breakpoints } from '@edx/paragon';
import { AppContext } from '@edx/frontend-platform/react';
import { Context as ResponsiveContext } from 'react-responsive';

import '../../../__mocks__/reactResponsive.mock';

import { CourseEnrollments } from '../CourseEnrollments';

const mockStore = configureMockStore([thunk]);

describe('<CourseEnrollments />', () => {
  const mockFetchCourseEnrollments = jest.fn();
  const mockClearCourseEnrollments = jest.fn();
  const mockModifyIsMarkCourseCompleteSuccess = jest.fn();
  const initialProps = {
    courseRuns: {
      in_progress: [],
      upcoming: [],
      completed: [],
    },
    isLoading: false,
    error: null,
    sidebarComponent: <div className="sidebar-example" />,
    fetchCourseEnrollments: mockFetchCourseEnrollments,
    clearCourseEnrollments: mockClearCourseEnrollments,
    isMarkCourseCompleteSuccess: false,
    modifyIsMarkCourseCompleteSuccess: mockModifyIsMarkCourseCompleteSuccess,
  };

  describe('renders course enrollments correctly', () => {
    it('with no course enrollments', () => {
      const pageContext = {};
      const wrapper = mount((
        <AppContext.Provider value={{ pageContext }}>
          <CourseEnrollments {...initialProps} />
        </AppContext.Provider>
      ));
      expect(wrapper.exists('.course-section')).toBeFalsy();
    });

    it('with valid course enrollments', () => {
      const sampleCourseRun = {
        courseRunId: 'course-v1:edX+DemoX+Demo_Course',
        courseRunStatus: 'completed',
        linkToCourse: 'https://edx.org/',
        title: 'edX Demonstration Course',
        notifications: [],
        startDate: '2017-02-05T05:00:00Z',
        endDate: '2018-08-18T05:00:00Z',
        hasEmailsEnabled: true,
      };
      const courseRuns = {
        in_progress: [{
          ...sampleCourseRun,
          courseRunId: 'course-v1:edX+DemoX+Demo_Course_2',
          courseRunStatus: 'in_progress',
          title: 'edX Demonstration Course 2',
          notifications: [{
            name: 'Assignment 1',
            url: 'https://edx.org/',
            date: '2019-05-31T07:50:00Z',
          }],
          microMastersTitle: 'Example MicroMasters Program',
        }],
        upcoming: [],
        completed: [sampleCourseRun],
      };
      const pageContext = {
        programUUID: 'test-program-uuid',
      };
      const store = mockStore({
        userAccount: {
          username: 'edx',
        },
        emailSettings: {
          loading: false,
          error: null,
          data: null,
        },
      });
      const wrapper = mount((
        <Provider store={store}>
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments
              {...initialProps}
              courseRuns={courseRuns}
            />
          </AppContext.Provider>
        </Provider>
      ));

      expect(wrapper.html()).not.toBeNull();
      expect(wrapper.find('.course-section').length).toEqual(2);
      expect(wrapper.find('.course-section').first().find('.course').length).toEqual(1);
      expect(wrapper.find('.course-section').last().find('.course').length).toEqual(1);
    });

    it('with error', () => {
      const pageContext = {};
      const tree = renderer
        .create((
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments
              {...initialProps}
              error={new Error('Network Error')}
            />
          </AppContext.Provider>
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with loading', () => {
      const pageContext = {};
      const tree = renderer
        .create((
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments
              {...initialProps}
              isLoading
            />
          </AppContext.Provider>
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with mark course as complete success status alert', () => {
      const pageContext = {};
      const tree = renderer
        .create((
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments
              {...initialProps}
              isMarkCourseCompleteSuccess
            />
          </AppContext.Provider>
        ))
        .toJSON();
      // TODO: why is this an array?
      expect(tree[0]).toMatchSnapshot();
    });
  });

  describe('sidebar', () => {
    let wrapper;

    it('is not shown at screen widths greater than or equal to large breakpoint', () => {
      const pageContext = {};
      wrapper = mount((
        <ResponsiveContext.Provider value={{ width: breakpoints.large.minWidth }}>
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments {...initialProps} />
          </AppContext.Provider>
        </ResponsiveContext.Provider>
      ));
      expect(wrapper.find('.sidebar-example').exists()).toBeFalsy();
    });

    it('is shown at screen widths less than large breakpoint', () => {
      const pageContext = {};
      wrapper = mount((
        <ResponsiveContext.Provider value={{ width: breakpoints.small.minWidth }}>
          <AppContext.Provider value={{ pageContext }}>
            <CourseEnrollments {...initialProps} />
          </AppContext.Provider>
        </ResponsiveContext.Provider>
      ));
      expect(wrapper.find('.sidebar-example').exists()).toBeTruthy();
    });
  });

  describe('calls appropriate fetch method depending on page type', () => {
    beforeEach(() => {
      mockFetchCourseEnrollments.mockReset();
      mockClearCourseEnrollments.mockReset();
    });

    it('for program page', () => {
      const programUUID = 'test-program-uuid';
      const pageContext = { programUUID };
      mount((
        <AppContext.Provider value={{ pageContext }}>
          <CourseEnrollments {...initialProps} />
        </AppContext.Provider>
      ));
      expect(mockFetchCourseEnrollments.mock.calls.length).toEqual(1);
      expect(mockFetchCourseEnrollments).toBeCalledWith({
        programUUID,
      });
    });
  });

  it('properly closes mark course as complete success status alert', () => {
    const pageContext = {
      enterpriseUUID: 'test-enterprise-uuid',
    };
    const wrapper = mount((
      <AppContext.Provider value={{ pageContext }}>
        <CourseEnrollments
          {...initialProps}
          isMarkCourseCompleteSuccess
        />
      </AppContext.Provider>
    ));
    wrapper.find('.alert-success .btn').simulate('click');
    expect(mockModifyIsMarkCourseCompleteSuccess).toBeCalledTimes(1);
  });
});
