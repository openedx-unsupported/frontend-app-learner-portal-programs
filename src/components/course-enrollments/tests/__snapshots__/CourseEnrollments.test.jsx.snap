// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseEnrollments /> renders course enrollments correctly with error 1`] = `
<div
  className="fade alert-content alert alert-danger show"
  role="alert"
>
  <span
    className="pgn__icon alert-icon"
  >
    <svg
      aria-hidden={true}
      fill="none"
      focusable={false}
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        fill="currentColor"
      />
    </svg>
  </span>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      <div
        className="d-flex"
      >
        An error occurred while retrieving your course enrollments. Please try again.
      </div>
    </div>
  </div>
</div>
`;

exports[`<CourseEnrollments /> renders course enrollments correctly with loading 1`] = `
<div
  className="loading-spinner d-flex justify-content-center align-items-center"
>
  <div
    className="spinner-border text-primary"
    role="status"
  >
    <span
      className="sr-only"
    >
      loading course enrollments
    </span>
  </div>
</div>
`;

exports[`<CourseEnrollments /> renders course enrollments correctly with mark course as complete success status alert 1`] = `
<div
  className="fade alert-content alert alert-success show"
  role="alert"
>
  <span
    className="pgn__icon alert-icon"
  >
    <svg
      aria-hidden={true}
      fill="none"
      focusable={false}
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        fill="currentColor"
      />
    </svg>
  </span>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      <div
        className="d-flex"
      >
        Your course was marked as complete.
      </div>
    </div>
    <div
      className="pgn__alert-actions pgn__action-row"
    >
      <span
        className="pgn__action-row-spacer"
      />
      <button
        className="btn btn-tertiary btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </div>
  </div>
</div>
`;
